ThreadSanitizer build fails on CI

`configure` fails at this point:

```
-- Performing Test OPENSSL_CORRECT_VERSION_NUMBER
-- Performing Test OPENSSL_CORRECT_VERSION_NUMBER - Failed
CMake Error at cmake/OpenSSLTests.cmake:82 (message):
  OpenSSL library version does not match headers
Call Stack (most recent call first):
  CMakeLists.txt:1069 (include)
-- Configuring incomplete, errors occurred!
See also "/zeek/build/CMakeFiles/CMakeOutput.log".
See also "/zeek/build/CMakeFiles/CMakeError.log".
Also, before re-running configure, consider cleaning the cache (removing the build directory) via `make distclean`
```

The error message is a red herring though, as the OpenSSL versions are fine. CMakeError.log shows the following:

```
/usr/bin/cc -DOPENSSL_CORRECT_VERSION_NUMBER -I/zeek/auxil/rapidjson/include -I/zeek/auxil/paraglob/include -I/zeek/auxil/highwayhash -I/zeek/build/auxil/broker/include -I/zeek/auxil/broker/include -I/zeek/build/auxil/binpac/lib -I/zeek/auxil/binpac/lib -Wall -Wno-unused -funsigned-char -g -DDEBUG -DBRO_DEBUG -fsanitize=thread -fno-omit-frame-pointer -fno-optimize-sibling-calls -O1  -fPIE -o CMakeFiles/cmTC_3b97f.dir/src.c.o -c /zeek/build/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_3b97f
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_3b97f.dir/link.txt --verbose=1
/usr/bin/cc  -Wall -Wno-unused -funsigned-char -g -DDEBUG -DBRO_DEBUG -fsanitize=thread -fno-omit-frame-pointer -fno-optimize-sibling-calls -O1  CMakeFiles/cmTC_3b97f.dir/src.c.o -o cmTC_3b97f  /usr/lib/x86_64-linux-gnu/libssl.so /usr/lib/x86_64-linux-gnu/libcrypto.so -ldl 
gmake[1]: Leaving directory '/zeek/build/CMakeFiles/CMakeTmp'


...and run output:
FATAL: ThreadSanitizer: unexpected memory mapping 0x5bb456972000-0x5bb456973000

Return value: 66
Source file was:

        #include <stdio.h>
        #include <openssl/opensslv.h>
        #include <openssl/crypto.h>
        int main() {
            printf("-- OpenSSL Library version: %s\n", SSLeay_version(SSLEAY_VERSION));
            printf("-- OpenSSL Header version: %s\n", OPENSSL_VERSION_TEXT);
            if (SSLeay() == OPENSSL_VERSION_NUMBER) {
                return 0;
            }
            return -1;
        }
```

I've tried switching from the default GCC 11.4 to both GCC 12.3 and Clang 15 with no success. This is also fine on a local version of our CI docker image. It only fails on Cirrus. I've tried forcing a rebuild of their image in case it was some transient issue with a package but it's still failing. I'm able to use the terminal access mode on Cirrus to reproduce it manually.

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -384,13 +384,15 @@ freebsd13_task:
 asan_sanitizer_task:
   container:
     # Just uses a recent/common distro to run memory error/leak checks.
-    dockerfile: ci/ubuntu-22.04/Dockerfile
+    dockerfile: ci/ubuntu-24.04/Dockerfile
     << : *RESOURCES_TEMPLATE
 
   << : *CI_TEMPLATE
   test_fuzzers_script: ./ci/test-fuzzers.sh
   coverage_script: ./ci/upload-coverage.sh
   env:
+    CC: clang-18
+    CXX: clang++-18
     CXXFLAGS: -DZEEK_DICT_DEBUG
     ZEEK_CI_CONFIGURE_FLAGS: *ASAN_SANITIZER_CONFIG
     ASAN_OPTIONS: detect_leaks=1:detect_odr_violation=0
@@ -398,33 +400,37 @@ asan_sanitizer_task:
 ubsan_sanitizer_task:
   container:
     # Just uses a recent/common distro to run undefined behavior checks.
-    dockerfile: ci/ubuntu-22.04/Dockerfile
+    dockerfile: ci/ubuntu-24.04/Dockerfile
     << : *RESOURCES_TEMPLATE
 
   << : *CI_TEMPLATE
   << : *SKIP_TASK_ON_PR
   test_fuzzers_script: ./ci/test-fuzzers.sh
   env:
+    CC: clang-18
+    CXX: clang++-18
     CXXFLAGS: -DZEEK_DICT_DEBUG
     ZEEK_CI_CONFIGURE_FLAGS: *UBSAN_SANITIZER_CONFIG
     ZEEK_TAILORED_UB_CHECKS: 1
     UBSAN_OPTIONS: print_stacktrace=1
 
-# tsan_sanitizer_task:
-#   container:
-#     # Just uses a recent/common distro to run memory error/leak checks.
-#     dockerfile: ci/ubuntu-22.04/Dockerfile
-#     << : *RESOURCES_TEMPLATE
+tsan_sanitizer_task:
+  container:
+    # Just uses a recent/common distro to run memory error/leak checks.
+    dockerfile: ci/ubuntu-24.04/Dockerfile
+    << : *RESOURCES_TEMPLATE
 
-#   << : *CI_TEMPLATE
-#   << : *SKIP_TASK_ON_PR
-#   env:
-#     ZEEK_CI_CONFIGURE_FLAGS: *TSAN_SANITIZER_CONFIG
-#     ZEEK_CI_DISABLE_SCRIPT_PROFILING: 1
-#     # If this is defined directly in the environment, configure fails to find
-#     # OpenSSL. Instead we define it with a different name and then give it
-#     # the correct name in the testing scripts.
-#     ZEEK_TSAN_OPTIONS: suppressions=/zeek/ci/tsan_suppressions.txt
+  << : *CI_TEMPLATE
+  << : *SKIP_TASK_ON_PR
+  env:
+    CC: clang-18
+    CXX: clang++-18
+    ZEEK_CI_CONFIGURE_FLAGS: *TSAN_SANITIZER_CONFIG
+    ZEEK_CI_DISABLE_SCRIPT_PROFILING: 1
+    # If this is defined directly in the environment, configure fails to find
+    # OpenSSL. Instead we define it with a different name and then give it
+    # the correct name in the testing scripts.
+    ZEEK_TSAN_OPTIONS: suppressions=/zeek/ci/tsan_suppressions.txt
 
 windows_task:
   # 2 hour timeout just for potential of building Docker image taking a while

--- a/ci/ubuntu-24.04/Dockerfile
+++ b/ci/ubuntu-24.04/Dockerfile
@@ -11,6 +11,8 @@ RUN apt-get update && apt-get -y install \
     bison \
     bsdmainutils \
     ccache \
+    clang-18 \
+    clang++-18 \
     cmake \
     curl \
     flex \

